build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/JavaApiMemberWriter.java

```diff
@@ -80,6 +83,16 @@ public void writeMethod(MethodMember method) {
             method.getExceptions().toArray(new String[0]));
         writeMethodAnnotations(mv, method.getAnnotations());
         writeMethodAnnotations(mv, method.getParameterAnnotations());
+
+        // Without this parameter annotations for non-static inner class constructors would
+        // end up having one more parameter than they should, because the first parameter
+        // (pointing to `this` of the enclosing type) should not be listed.
+        if (method.getSignature() != null) {
+            int parameterCount = SignatureParameterCounter.countParameters(method.getSignature());
+            mv.visitAnnotableParameterCount(parameterCount, true);
+            mv.visitAnnotableParameterCount(parameterCount, false);
+        }
+
```

Comment:
💭 This is the main change. We are now explicitly telling ASM how many annotable parameters there are instead of trusting it to figure it out based on the method descriptor.